Fixed #32915 -- Didn't catch ImportErrors unless it's the settings module.

bcail opened the pull request:
…le that can't be imported

Based on work by Rohith PR and others in PR #14622

bcail committed:
Fixed #32915 - Don't catch ImportErrors unless it's the settings module that can't be imported

Based on work by Rohith PR and others in PR #14622

bcail committed:
run black

bcail committed:
fix flake8 errors

felixxm commented on django/core/management/__init__.py:
```diff
@@ -382,7 +382,9 @@ def execute(self):
         except ImproperlyConfigured as exc:
             self.settings_exception = exc
-        except ImportError as exc:
+            # The following commands can be run without a valid settings file configured
+            if subcommand not in {"startproject", "startapp", "makemessages"}:
+                sys.stderr.write(str(exc) + "\n")
```
https://github.com/django/django/pull/14622#discussion_r797244647 is not addressed.

Also, we now set `self.settings_exception = exc` twice.

bcail replied:
@felixxm thanks for the review. I made the other changes you suggested, and rebased the branch. Do you have suggestions for how to not write the exception message for these three subcommands, without hard-coding them in the set?

jacobtylerwalls replied:
How about a boolean attribute on `BaseCommand` akin to `requires_migrations_checks` like `requires_settings`?

bcail replied:
Thanks. It looks like `subcommand` is still just a string, though, so it might not be able to check a boolean attribute on `BaseCommand`?

jacobtylerwalls replied:
You have the command class by the end of the function, where you could assign the result of fetch_command(). Is there a way to delay the writing to stderr until that is done?

bcail replied:
I pushed up a commit adding the `requires_settings` attribute... is that the direction you were thinking?

bcail replied:
@jacobtylerwalls any thoughts on my latest commit here?

felixxm commented on tests/settings_tests/tests.py:
```diff
@@ -382,6 +382,23 @@ def test_use_deprecated_pytz_deprecation(self):
+            Settings("fake_settings_module")
+
+    def test_unable_to_import_a_random_module(self):
+        def mock_import_module(_):
+            raise ModuleNotFoundError(
+                "No module named 'fake_module'", name="fake_module"
+            )
+
```
This can be handled with the `side_effect`:
```python
        exc = ModuleNotFoundError("No module named 'fake_module'", name="fake_module")
        with mock.patch("importlib.import_module", side_effect=exc):
            with self.assertRaisesMessage(ImportError, "No module named 'fake_module'"):
                Settings("fake_settings_module")
```

felixxm commented on tests/settings_tests/tests.py:
```diff
@@ -382,6 +382,23 @@ def test_use_deprecated_pytz_deprecation(self):
             holder.configure(USE_DEPRECATED_PYTZ=True)
 
+    def test_unable_to_import_settings_module(self):
+        with self.assertRaisesMessage(
+            ImproperlyConfigured,
+            "Settings module fake_settings_module could not be imported",
+        ):
```
```suggestion
        msg = "Settings module fake_settings_module could not be imported"
        with self.assertRaisesMessage(ImproperlyConfigured, msg):
```

felixxm reviewed (commented):
@bcail Thanks for this patch :+1: I left small comments.

bcail committed:
PR review

bcail commented:
@felixxm I rebased this branch, and it looks like one job failed - but it doesn't seem related to this PR. Is there a way to re-run the job?

Also, can you let me know what kind of change you're looking for in your comment above?

Thanks.

bcail committed:
add test for case in the ticket

bcail committed:
fix linters

bcail committed:
add requires_settings attribute

bcail committed:
remove unnecessary call to settings.INSTALLED_APPS

bcail committed:
add a test for a command that doesn't require settings

sarahboyce reviewed (commented):
I still get the error described in the ticket when testing this patch

bcail commented:
Thanks for taking a look, @sarahboyce. The difference from this branch is that it adds a note that the settings module couldn't be imported:
```
usage: manage.py runserver [-h] [--ipv6] [--nothreading] [--noreload]
                           [--version] [--settings SETTINGS]
                           [--pythonpath PYTHONPATH] [--no-color]
                           [--force-color] [--skip-checks]
                           [addrport]
manage.py runserver: error: unrecognized arguments: --nostatic
```
becomes:
```
Settings module settings could not be imported
usage: manage.py runserver [-h] [--ipv6] [--nothreading] [--noreload]
                           [--version] [--settings SETTINGS]
                           [--pythonpath PYTHONPATH] [--no-color]
                           [--force-color] [--skip-checks]
                           [addrport]
manage.py runserver: error: unrecognized arguments: --nostatic
```

Maybe if settings are needed, and not found, any subsequent exceptions should just be suppressed completely and only the settings error should be displayed?

bcail committed:
quit immediately if settings module couldn't be imported

bcail committed:
immediately quit if settings module wasn't specified at all

bcail commented:
@sarahboyce I updated this PR to exit immediately if the settings module can't be imported. I also changed it to exit immediately if the settings module name isn't set.

If this looks OK to you, I can work on the docs.

sarahboyce commented on django/conf/__init__.py:
```diff
@@ -163,7 +163,15 @@ def __init__(self, settings_module):
+            # If settings module cannot be imported, treat it as a configuration error.
+            if exc.name == self.SETTINGS_MODULE:
+                raise ImproperlyConfigured(
+                    f"Settings module {self.SETTINGS_MODULE} could not be imported"
```
```suggestion
                    f"Settings module '{self.SETTINGS_MODULE}' could not be imported."
```

sarahboyce commented on django/core/management/__init__.py:
```diff
@@ -254,14 +254,12 @@ def fetch_command(self, subcommand):
-                # informed about it.
-                settings.INSTALLED_APPS
+            if os.environ.get("DJANGO_SETTINGS_MODULE") and self.settings_exception:
+                sys.stderr.write(str(self.settings_exception) + "\n")
```
```suggestion
                sys.stderr.write(self.style.NOTICE(str(self.settings_exception) + "\n"))
```
And have the init set `self.style = color_style()` and update the others using a style

sarahboyce commented on django/core/management/base.py:
```diff
@@ -260,6 +260,7 @@ class BaseCommand:
     _called_from_command_line = False
     output_transaction = False  # Whether to wrap the output in a "BEGIN; COMMIT;"
     requires_migrations_checks = False
+    requires_settings = True
```
This probably should be False by default so custom management commands opt-in to the new behavior :thinking:

sarahboyce commented on django/core/management/commands/makemessages.py:
```diff
@@ -208,6 +208,7 @@ class Command(BaseCommand):
     translatable_file_class = TranslatableFile
     build_file_class = BuildFile
 
+    requires_settings = False
```
We should probably have a test which loops through all settings with this as False and another with this as True, and asserts based off bad settings (in a subTest)

jacobtylerwalls replied:
Good idea, I'll tweak `test_commands_with_invalid_settings`.

jacobtylerwalls replied:
Okay, I pushed something for this. It's a little bit of an "X is X" test, but maybe it has some value.

jacobtylerwalls replied:
EDIT: I don't have permissions to push to this PR, so I'll open a successor PR.

sarahboyce reviewed (commented):
Thank you for the updates @bcail - I like it :+1: added a few more thoughts

bcail committed:
Update settings error message

Co-authored-by: Sarah Boyce <42296566+sarahboyce@users.noreply.github.com>

bcail committed:
Style the settings error message

Co-authored-by: Sarah Boyce <42296566+sarahboyce@users.noreply.github.com>

bcail committed:
color style

bcail committed:
style another settings exc notice

bcail commented:
For the new `requires_settings` attribute - if I'm reading [get_commands](https://github.com/django/django/blob/main/django/core/management/__init__.py#L53) and [fetch_command](https://github.com/django/django/blob/main/django/core/management/__init__.py#L246) correctly, I think it will only be core commands that will get to the point of the `requires_settings` check. Any custom commands shouldn't be found at all (because settings aren't configured), so they'll go into the `except KeyError` block.

So I think having the default be `True` should be OK?

bcail commented on django/core/management/__init__.py:
```diff
@@ -254,14 +254,12 @@ def fetch_command(self, subcommand):
             elif not settings.configured:
-                sys.stderr.write("No Django settings specified.\n")
+                sys.stderr.write(self.style.NOTICE("No Django settings specified.\n"))
+                sys.exit(1)
```
I'm thinking maybe I should remove this `sys.exit(1)` (and the one 3 lines up), and let it check for close matches before exiting. So that we could get this:
```
$ python3 ./manage.py startap
Settings module 'settings' could not be imported.
Unknown command: 'startap'. Did you mean startapp?
Type 'manage.py help' for usage.
```

bcail replied:
No, the `requires_settings` attribute is only checked below, if the command is found despite there being a settings exception.

jacobtylerwalls commented:
I applied review feedback in a new PR (#21157), since I didn't appear to have permissions to write to this one. Reviews welcomed there!
